### PR TITLE
Add required dependencies to set up Azure ccache in workflows

### PR DESCRIPTION
**Pull request by Eliasj42**

### CODE REVIEW

File: dockerfiles/cpubuilder_ubuntu_jammy_x86_64.Dockerfile
Comment (ScottTodd) [inline dockerfiles/cpubuilder_ubuntu_jammy_x86_64.Dockerfile:1]: Please rebase your branch on the latest code. There are merge conflicts:
image

Comment (ScottTodd) [inline dockerfiles/cpubuilder_ubuntu_jammy_x86_64.Dockerfile:1]: Please add commit authorship information and resolve the DCO check: https://github.com/iree-org/base-docker-images/pull/7/checks?check_run_id=29708830851

Documentation for the DCO check is here: https://iree.dev/developers/general/contributing/#developer-certificate-of-origin . Adding the "Signed-off-by" text is simplest (https://iree.dev/developers/general/contributing/#adding-signed-off-by-to-commits), but SSH or GPG signing is seamless and will benefit all commits you sent to GitHub, so I consider it worth the extra effort.

File: dockerfiles/cpubuilder_ubuntu_jammy_x86_64.Dockerfile [hunk 1]
```diff
@@ -18,6 +18,13 @@ RUN apt install -y \
         libssl-dev libxml2-dev libcapstone-dev libtbb-dev libzstd-dev \
         llvm-dev pkg-config
 # Cleanup
+RUN curl -sSL -O https://packages.microsoft.com/config/ubuntu/22.04/packages-microsoft-prod.deb
```
Comment (ScottTodd) [inline dockerfiles/cpubuilder_ubuntu_jammy_x86_64.Dockerfile:21]: These are not cleanup steps, so they shouldn't go in this section of code.

I'd structure this like so:
```diff
######## Apt packages ########
RUN apt update && \
    apt install -y wget git unzip curl gnupg2 lsb-release
# Python 3.
RUN apt install -y python3.11-dev python3.11-venv python3-pip && \
    update-alternatives --install /usr/local/bin/python python /usr/bin/python3.11 3 && \
    update-alternatives --install /usr/local/bin/python3 python3 /usr/bin/python3.11 3
# Toolchains and build dependencies.
RUN apt install -y \
        clang-14 lld-14 \
        gcc-9 g++-9 \
-       ccache ninja-build libssl-dev libxml2-dev libcapstone-dev libtbb-dev \
+       ninja-build libssl-dev libxml2-dev libcapstone-dev libtbb-dev \
        libzstd-dev llvm-dev pkg-config
# Recent compiler tools for build configurations like ASan/TSan.
#   * See https://apt.llvm.org/ for context on the apt commands.
ARG LLVM_VERSION=19
RUN echo "deb http://apt.llvm.org/jammy/ llvm-toolchain-jammy-${LLVM_VERSION} main" >> /etc/apt/sources.list && \
    curl https://apt.llvm.org/llvm-snapshot.gpg.key | gpg --dearmor > /etc/apt/trusted.gpg.d/llvm-snapshot.gpg && \
    apt update && \
    apt install -y clang-${LLVM_VERSION} lld-${LLVM_VERSION}
+ # Virtual file system adapter for Azure Blob storage, used for remote cache.
+ RUN curl -sSL -O https://packages.microsoft.com/config/ubuntu/22.04/packages-microsoft-prod.deb
+ RUN dpkg -i packages-microsoft-prod.deb
+ RUN rm packages-microsoft-prod.deb
+ RUN apt-get update && \
+     apt-get install -y fuse3 blobfuse2
# Cleanup.
RUN apt clean && rm -rf /var/lib/apt/lists/*

+ ######## CCache ########
+ WORKDIR /install-ccache
+ COPY build_tools/install_ccache.sh ./
+ RUN ./install_ccache.sh "4.10.2" && rm -rf /install-ccache

######## CMake ########
WORKDIR /install-cmake
ENV CMAKE_VERSION="3.23.2"
COPY build_tools/install_cmake.sh ./
RUN ./install_cmake.sh "${CMAKE_VERSION}" && rm -rf /install-cmake
```

(Untested, please check that those changes work)

File: dockerfiles/cpubuilder_ubuntu_jammy_x86_64.Dockerfile
Comment (ScottTodd) [inline dockerfiles/cpubuilder_ubuntu_jammy_x86_64.Dockerfile:1]: Until we decide otherwise, please mirror any changes from this file to https://github.com/iree-org/base-docker-images/blob/main/dockerfiles/cpubuilder_ubuntu_jammy_ghr_x86_64.Dockerfile. The files are the same except for the first few and last few lines.

**Review by ScottTodd — CHANGES_REQUESTED** [review]

File: dockerfiles/cpubuilder_ubuntu_jammy_x86_64.Dockerfile [hunk 2]
```diff
@@ -21,22 +22,33 @@ RUN echo "deb http://apt.llvm.org/jammy/ llvm-toolchain-jammy-${LLVM_VERSION} ma
...
 ######## Build toolchain configuration ########
-# Setup symlinks and alternatives then default to using clang-14.
-# This can be overriden to gcc or another clang version as needed.
+# Setup symlinks and alternatives.
 RUN ln -s /usr/bin/lld-14 /usr/bin/lld && \
     ln -s /usr/bin/ld.lld-14 /usr/bin/ld.lld && \
     ln -s /usr/bin/clang-14 /usr/bin/clang && \
     ln -s /usr/bin/clang++-14 /usr/bin/clang++
+# Default to using clang. This can be overriden to gcc as desired.
 ENV CC=clang
```
Comment (ScottTodd) [inline dockerfiles/cpubuilder_ubuntu_jammy_x86_64.Dockerfile:52]: Some diffs slipped in here while rebasing. Can you clean those up? (Generally, doing a self-review on your PR can help catch these)

File: dockerfiles/cpubuilder_ubuntu_jammy_x86_64.Dockerfile [hunk 3]
```diff
@@ -1,4 +1,4 @@
-# Stock Ubuntu Jammy (22.04) with IREE build dependencies.
+# Stock Ubuntu with IREE build deps.
 FROM ubuntu:jammy
```
Comment (ScottTodd) [inline dockerfiles/cpubuilder_ubuntu_jammy_x86_64.Dockerfile:2]: Revert unrelated comment changes please

File: dockerfiles/cpubuilder_ubuntu_jammy_x86_64.Dockerfile
Comment (ScottTodd) [reply dockerfiles/cpubuilder_ubuntu_jammy_x86_64.Dockerfile:1]: Can you mirror these changes to that file?

**Review by ScottTodd — COMMENTED** [review]
Thanks. Just a few nits then This should be ready I think.

File: dockerfiles/cpubuilder_ubuntu_jammy_ghr_x86_64.Dockerfile [hunk 4]
```diff
@@ -13,6 +13,7 @@ RUN apt install -y \
         clang-14 lld-14 \
         gcc-9 g++-9 \
         ccache ninja-build libssl-dev libxml2-dev libcapstone-dev libtbb-dev \
+        ninja-build libssl-dev libxml2-dev libcapstone-dev libtbb-dev \
```
Comment (Eliasj42) [inline dockerfiles/cpubuilder_ubuntu_jammy_ghr_x86_64.Dockerfile:16]: is this supposed to be added?

Comment (ScottTodd) [reply dockerfiles/cpubuilder_ubuntu_jammy_ghr_x86_64.Dockerfile:16]: Oh, I meant to just remove "ccache" from that line, since the Dockerfile installs a specific version later using `install_ccache.sh`.

File: dockerfiles/cpubuilder_ubuntu_jammy_ghr_x86_64.Dockerfile [hunk 5]
```diff
@@ -37,6 +49,7 @@ RUN ln -s /usr/bin/lld-14 /usr/bin/lld && \
     ln -s /usr/bin/ld.lld-14 /usr/bin/ld.lld && \
     ln -s /usr/bin/clang-14 /usr/bin/clang && \
     ln -s /usr/bin/clang++-14 /usr/bin/clang++
+# Default to using clang. This can be overriden to gcc as desired.
```
Comment (ScottTodd) [inline dockerfiles/cpubuilder_ubuntu_jammy_ghr_x86_64.Dockerfile:52]: Drop extra comment. That's already covered above:
```
# Setup symlinks and alternatives then default to using clang-14.
# This can be overriden to gcc or another clang version as needed.
```

(I guess I had a typo in there, - should be "overridden" could fix or ignore)

File: dockerfiles/cpubuilder_ubuntu_jammy_ghr_x86_64.Dockerfile (hunk 4 above)
Comment (Eliasj42) [reply dockerfiles/cpubuilder_ubuntu_jammy_ghr_x86_64.Dockerfile:16]: sorry, I'm a little confused, ccache isnt in that line

Comment (ScottTodd) [reply dockerfiles/cpubuilder_ubuntu_jammy_ghr_x86_64.Dockerfile:16]: Replace
```
        ccache ninja-build libssl-dev libxml2-dev libcapstone-dev libtbb-dev \
```
with
```dockerfile
        ninja-build libssl-dev libxml2-dev libcapstone-dev libtbb-dev \
```

Right now you have both lines, so `ninja-build` is listed twice, for example
```
        ccache ninja-build libssl-dev libxml2-dev libcapstone-dev libtbb-dev \
        ninja-build libssl-dev libxml2-dev libcapstone-dev libtbb-dev \
```

If that gets too confusing or the diffs are too noisy, we can also put these one per line to make such diffs easier to read:
```
        ninja-build \
        libssl-dev \
        libxml2-dev \
        libcapstone-dev \
        libtbb-dev \
```

Comment (Eliasj42) [reply dockerfiles/cpubuilder_ubuntu_jammy_ghr_x86_64.Dockerfile:16]: oh i see mb

**Review by ScottTodd — APPROVED** [review]

File: dockerfiles/cpubuilder_ubuntu_jammy_ghr_x86_64.Dockerfile (hunk 4 above)
Comment (ScottTodd) [reply dockerfiles/cpubuilder_ubuntu_jammy_ghr_x86_64.Dockerfile:15]: Yep! I'll merge now if that's okay.

**Comment by ScottTodd** [comment]
Errr, well technically these should be your own name and email:
```
Signed-off-by: Ubuntu <eljoseph@Elias-Dev-2.tqhyce5jp0cu3oroh4dbeksqrg.xx.internal.cloudapp.net>
Signed-off-by: Ubuntu <eljoseph@Elias-Dev-2.tqhyce5jp0cu3oroh4dbeksqrg.xx.internal.cloudapp.net>
Signed-off-by: Ubuntu <eljoseph@Elias-Dev-2.tqhyce5jp0cu3oroh4dbeksqrg.xx.internal.cloudapp.net>
```

This should set it:
```
git config --global user.email name@company.ccc
git config --global user.name 'First Last'
```

then commit as you would normally

**Comment by ScottTodd** [comment]
Close enough 😄 . Build kicked off at https://github.com/iree-org/base-docker-images/actions/runs/10728890623 automatically. Once that finishes (assuming successful), you can grab the new SHA from https://github.com/iree-org/base-docker-images/pkgs/container/cpubuilder_ubuntu_jammy_x86_64
